File: ios/Saga/Sources/Models/Events/EventRole.swift
import Foundation

// MARK: - Event Role

/// A role that can be assigned to attendees at an event
struct EventRole: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let eventId: String
    let catalogRoleId: String?
    let name: String
    let description: String?
    let maxSlots: Int
    let filledSlots: Int

    enum CodingKeys: String, CodingKey {
        case id
        case eventId = "event_id"
        case catalogRoleId = "catalog_role_id"
        case name
        case description
        case maxSlots = "max_slots"
        case filledSlots = "filled_slots"
    }

    /// Whether this role has all slots filled
    var isFull: Bool {
        filledSlots >= maxSlots
    }

    /// Number of spots remaining
    var spotsLeft: Int {
        max(0, maxSlots - filledSlots)
    }

    /// Availability text like "2/5 filled"
    var availabilityText: String {
        "\(filledSlots)/\(maxSlots) filled"
    }
}

// MARK: - Role Assignment

/// An assignment of a user to a role
struct RoleAssignment: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let eventId: String
    let roleId: String
    let userId: String
    let note: String?
    let status: RoleAssignmentStatus
    let assignedOn: Date

    enum CodingKeys: String, CodingKey {
        case id
        case eventId = "event_id"
        case roleId = "role_id"
        case userId = "user_id"
        case note
        case status
        case assignedOn = "assigned_on"
    }
}

// MARK: - Role Assignment Status

enum RoleAssignmentStatus: String, Codable, Sendable, CaseIterable {
    case assigned
    case confirmed
    case declined

    var displayName: String {
        switch self {
        case .assigned: return "Assigned"
        case .confirmed: return "Confirmed"
        case .declined: return "Declined"
        }
    }
}

// MARK: - Event Role With Assignments

/// A role bundled with its assignments
struct EventRoleWithAssignments: Codable, Sendable {
    let role: EventRole
    let assignments: [RoleAssignment]
    let isFull: Bool
    let spotsLeft: Int

    enum CodingKeys: String, CodingKey {
        case role
        case assignments
        case isFull = "is_full"
        case spotsLeft = "spots_left"
    }
}

// MARK: - Event Roles Overview

/// Overview of all roles for an event
struct EventRolesOverview: Codable, Sendable {
    let data: [EventRoleWithAssignments]
}

// MARK: - Role Suggestion

/// A suggested role based on user interests
struct RoleSuggestion: Codable, Identifiable, Sendable {
    let role: RoleCatalogEntry
    let reason: String
    let usageCount: Int

    var id: String { role.id }

    enum CodingKeys: String, CodingKey {
        case role
        case reason
        case usageCount = "usage_count"
    }
}

// MARK: - Role Catalog Entry

/// A reusable role from the catalog
struct RoleCatalogEntry: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let scopeType: String
    let scopeId: String
    let roleType: String
    let name: String
    let description: String?
    let icon: String?
    let isActive: Bool
    let createdBy: String?
    let createdOn: Date

    enum CodingKeys: String, CodingKey {
        case id
        case scopeType = "scope_type"
        case scopeId = "scope_id"
        case roleType = "role_type"
        case name
        case description
        case icon
        case isActive = "is_active"
        case createdBy = "created_by"
        case createdOn = "created_on"
    }
}

// MARK: - Request Types

struct CreateEventRoleRequest: Codable, Sendable {
    let catalogRoleId: String?
    let name: String
    let description: String?
    let maxSlots: Int

    enum CodingKeys: String, CodingKey {
        case catalogRoleId = "catalog_role_id"
        case name
        case description
        case maxSlots = "max_slots"
    }

    init(name: String, description: String? = nil, maxSlots: Int = 1, catalogRoleId: String? = nil) {
        self.name = name
        self.description = description
        self.maxSlots = maxSlots
        self.catalogRoleId = catalogRoleId
    }
}

struct UpdateEventRoleRequest: Codable, Sendable {
    var name: String?
    var description: String?
    var maxSlots: Int?

    enum CodingKeys: String, CodingKey {
        case name
        case description
        case maxSlots = "max_slots"
    }
}

struct AssignRoleRequest: Codable, Sendable {
    let roleId: String
    let userId: String?
    let note: String?

    enum CodingKeys: String, CodingKey {
        case roleId = "role_id"
        case userId = "user_id"
        case note
    }

    init(roleId: String, userId: String? = nil, note: String? = nil) {
        self.roleId = roleId
        self.userId = userId
        self.note = note
    }
}

File: api/internal/model/event_role.go
package model

import "time"

// EventRole represents a role that attendees can fill at an event
type EventRole struct {
	ID          string    `json:"id"`
	EventID     string    `json:"event_id"`
	Name        string    `json:"name"`                  // e.g., "Dessert-bringer", "DJ", "Guest"
	Description *string   `json:"description,omitempty"` // e.g., "Bring something sweet to share"
	MaxSlots    int       `json:"max_slots"`             // Default 1, 0 = unlimited (for default Guest role)
	FilledSlots int       `json:"filled_slots"`          // Computed from assignments
	IsDefault   bool      `json:"is_default"`            // True for the default "Guest" role
	SortOrder   int       `json:"sort_order"`            // Display ordering
	CreatedBy   string    `json:"created_by"`            // Host who created this role
	CreatedOn   time.Time `json:"created_on"`
	UpdatedOn   time.Time `json:"updated_on"`
	// Optional: suggest this role to users with matching interests
	SuggestedInterests []string `json:"suggested_interests,omitempty"`
}

// DefaultMaxSlotsPerRole is the default number of slots for a role (1 person per role)
const DefaultMaxSlotsPerRole = 1

// EventRoleAssignment represents a user's assignment to a role
// A user can have multiple role assignments per event (e.g., DJ + bring lasagna + wash dishes)
type EventRoleAssignment struct {
	ID         string    `json:"id"`
	EventID    string    `json:"event_id"`
	RoleID     string    `json:"role_id"`
	UserID     string    `json:"user_id"`
	Note       *string   `json:"note,omitempty"` // User's note about their contribution
	Status     string    `json:"status"`         // pending, confirmed, cancelled
	AssignedOn time.Time `json:"assigned_on"`
	UpdatedOn  time.Time `json:"updated_on"`
	// Populated by joins
	RoleName *string `json:"role_name,omitempty"`
}

// UserEventRoles represents all roles a user has taken on for an event
type UserEventRoles struct {
	UserID      string                `json:"user_id"`
	EventID     string                `json:"event_id"`
	Assignments []EventRoleAssignment `json:"assignments"`
}

// EventRoleAssignmentStatus constants
const (
	RoleAssignmentStatusPending   = "pending"   // User expressed interest, awaiting confirmation
	RoleAssignmentStatusConfirmed = "confirmed" // User is confirmed in this role
	RoleAssignmentStatusCancelled = "cancelled" // User cancelled their assignment
)

// Note: EventAttendanceConfig fields are now part of Event struct in event.go

// EventRoleWithAssignments includes the role and its current assignments
type EventRoleWithAssignments struct {
	Role        EventRole             `json:"role"`
	Assignments []EventRoleAssignment `json:"assignments"`
	IsFull      bool                  `json:"is_full"`
	SpotsLeft   int                   `json:"spots_left"` // -1 if unlimited
}

// EventRolesOverview provides a summary of all roles for an event
type EventRolesOverview struct {
	EventID        string                     `json:"event_id"`
	TotalSlots     int                        `json:"total_slots"`     // Total role slots available (-1 if unlimited)
	FilledSlots    int                        `json:"filled_slots"`    // Currently filled slots
	TotalAttendees int                        `json:"total_attendees"` // Same as filled_slots (for compatibility)
	MaxAttendees   *int                       `json:"max_attendees,omitempty"`
	IsFull         bool                       `json:"is_full"`
	Roles          []EventRoleWithAssignments `json:"roles"`
}

// Role name constants
const (
	DefaultRoleName        = "Guest"
	DefaultRoleDescription = "Join us and have a great time!"
)

// Constraints
const (
	MaxRolesPerEvent     = 20
	MaxRoleNameLength    = 50
	MaxRoleDescLength    = 500
	MaxAssignmentNoteLen = 200
)

// CreateEventRoleRequest represents a request to create a role
type CreateEventRoleRequest struct {
	Name               string   `json:"name"`
	Description        *string  `json:"description,omitempty"`
	MaxSlots           int      `json:"max_slots,omitempty"` // 0 = unlimited
	SuggestedInterests []string `json:"suggested_interests,omitempty"`
}

// UpdateEventRoleRequest represents a request to update a role
type UpdateEventRoleRequest struct {
	Name               *string  `json:"name,omitempty"`
	Description        *string  `json:"description,omitempty"`
	MaxSlots           *int     `json:"max_slots,omitempty"`
	SuggestedInterests []string `json:"suggested_interests,omitempty"`
}

// AssignRoleRequest represents a request to assign oneself to a role
type AssignRoleRequest struct {
	RoleID string  `json:"role_id"`
	Note   *string `json:"note,omitempty"` // e.g., "I'll bring vegan brownies"
}

// UpdateAssignmentRequest represents a request to update an assignment
type UpdateAssignmentRequest struct {
	Note *string `json:"note,omitempty"`
}

// RoleSuggestion represents a suggested role based on user interests
type RoleSuggestion struct {
	Role            EventRole `json:"role"`
	MatchedInterest string    `json:"matched_interest"`
	Reason          string    `json:"reason"` // e.g., "You're interested in baking"
}
